### DISCUSSION TITLE
ice.js->model.js 迁移

### DISCUSSION THREAD

**Discussion by xmsz-stu** [web-infra-dev/modern.js#6128] [General]
原先有个项目是ice.js 2.x，由于框架太旧很多新包用不了，很多新技术也用不上，所以打算升级

考虑
- next.js 踩了一堆坑，很麻烦，而且很多特性用不上和用不了，放弃
- ice.js 3.x，中规中矩，而且构建速度极速下降，加速方案用的也是rspack，放弃
- modern.js， 和rspack同生态，类next.js，后续可扩展，选择

渐进式更新
因为是老项目，所以一次性改起来，不仅麻烦而且风险很大，所以用渐进式
- 主项目保留ice 2.x, 子项目一个一个替换成modern.js
- 主项目同时保留原有微前端方案，加上手动加载Garfish运行新的子项目
- 先跑一阵子
- 开始升级主项目

遇到的情况
- 因为之前本来按规范写的，所以业务代码逻辑一个不用改，不用改也就代表不会有问题
- store替换成reduck，写法变了一点不过问题不大，然后由于需要父子应用状态同步，所以主子应用都独立使用reduck，然后父应用把store传递给子应用
- auth，这个modern.js没有内置，所以只能先把auth的状态也放到store里
- unocss，rspack支持，没问题，但是有一些小坑，比如unocss更新了，但是热更新不会更新... 
- 热更新，有一些小问题，比如样式更新每次都会全部加载一遍页面的css,一些cdn包特别大也加载导致速度变慢
- 父子应用间通信，garfish.channel没问题，不过注意的这个是挂在window上，刚开始不一定拿的到
- 类型检查，这个是唯一改得最多的，因为之前ts版本也好，规范也好都不一样，所以很多改，麻烦不过都是小问题，比如 onChange(v: string) {} -> onChange(v) { a = v as string }，还有reduce要自己指定类型
- 构建，这个比较蛋疼，配置了一堆东西，没有很方便。比如部分包用cdn加载、子应用要构建出 xxx/index.html 文件...
- wasm 没问题
- worker 改成正常写法后没问题
- 路由，这个跳转方式还是变了一些，不过改起来还行

所有项目升级完后遇到的问题
- 由于unocss可能是全量扫描包，导致现在每个项目每次热更新至少要3s，而且机器会狂响
-  

下一步
- 当然是升级react,之前是17
...

**Comment by lanmingle** [comment]
umi ：先行者，现在不用看了，浪费时间
next.js app router ：很奈斯，就是缺少微前端的支持，路由自定义
next.js pages router：挺完善的，缺少 loading 等提供的实现，微前端可用模块联邦
ice.js ：有着 next.js pages router 的影子
modern.js：有着上面全部的影子

modern.js 这个名字不好记住，建议改一下，哈哈哈！

我觉得 modern.js 目前很好了（为啥这么少人用呢？应该是名字难记和有点拽），就是缺少了一下插件支持，如 ui框架、auth，路由部分不够完善（如插座，易用的路由事件，不是自己实现），脚手架不给可选配置，国际化支持等等。

现在我对 modern.js 做评估，已有一段时间了，库工程很不错，web工程还在测试中，基本上已满足，不行的话只能弃车保帅。

**Reply by xmsz-stu** [reply]
没错 我也觉得很奇怪 为什么用的人这么少？内部用就不说了，社区是什么情况。 虽然大部分人还是umi或者徒手搓vue/react，但也不至于这样
